### PR TITLE
swap: extend randomEncryption to plainOpen and ability to select cipher

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

Based on https://github.com/NixOS/nixpkgs/pull/25900 I expanded the configuration of swapDevices.*.randomEncryption to a set of options to enable users to choose a cipher etc.

```
randomEncryption = (true|false);
```

becomes
```
randomEncryption = {
  enable = (true|false);
  cipher = "aes-xts-plain64";
  source = "/dev/urandom";
};
```

Of course __this breaks all current user configurations__ when they already use randomEncryption.

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fpletz** [comment]
You could use `coerceTo` from #21875 for backwards-compatibility.

But LGTM otherwise. I like the change from LUKS to plain mode. :+1:

**Comment by abbradar** [comment]
Huh, I did the tests and intended to merge #24900 but postoned it to later today, it was a good decision ~_^

Let's try to use `coerceTo` and if it doesn't work the second option would be to rename this subtree to say just `encryption` and add an entry to `rename.nix`. I like the new option!

**Comment by gentoofreak** [comment]
OK, I'll try to change it to `coerce` later this day. But I'm not that experienced with the nix language yet, so maybe I will fail or do it wrong. :)

**Comment by abbradar** [comment]
@gentoofreak I'll be happy to help with this -- don't hesitate to ask!

**Comment by gentoofreak** [comment]
@abbradar please review and test: I hope I've done the `coercedTo` change right.

**Comment by abbradar** [comment]
Tested in local VM; it works for me. I don't see any immediate other problem so I'll merge in a day with minor fixes (adding `trace` to notify users of deprecated configuration way) unless anybody objects.

Thank you for your work!

**Comment by gentoofreak** [comment]
@abbradar any plans when this gets merged?

File: nixos/modules/config/swap.nix [hunk 1]
```diff
@@ -5,6 +5,52 @@ with lib;
...
+          WARNING #2: Do not use /dev/disk/by-uuid/… or /dev/disk/by-label/… as your swap device
+          when using randomEncryption as the UUIDs and labels will get erased on every boot when
+          the partition is encrypted. Best to use /dev/disk/by-partuuid/…
+        '';
+      };
+
+      cipher = mkOption {
+        default = "aes-xts-plain64";
+        example = "serpent-xts-plain64";
+        type = types.str;
```
Comment (0xABAB) [inline nixos/modules/config/swap.nix]: Isn't the set of ciphers enumerable and computable?

**Comment by 0xABAB** [comment]
This PRs version is a step in the right direction over what there is now, but my understanding is that assertions could be added to make it a true module. E.g. it's now possible to use a cipher like "wrongcipher". 

Additionally, I think operations involving devices are pretty scary to begin with, so why don't we just use swapfiles, which are not meaningfully slower, as a default. See https://serverfault.com/a/277842.

I am fine with optimized paths, but there aren't even tests for this. So, why should anyone trust you aren't bricking people's machines? (Even if it is 100% correct now, perhaps some other person comes along later and breaks it, causing yourself data-loss, just because you didn't write a test).

Speaking of tests, are the tests actually run somewhere with public output?

**Comment by gentoofreak** [comment]
Data loss in this case is not a real problem because this applies only to swap. Worst case if it does not work is that the swap device is not fully setup and therefore you have no swap.

Another problem that could come into consideration: unexperienced users could make there swap encryption less secure than the default without noticing. But to be honest I think unexperienced users will only set `enable = true` and leave the other options to their default values.

**Comment by 0xABAB** [comment]
@gentoofreak You didn't respond to all my feedback. (See the code comment.)

File: nixos/modules/config/swap.nix (hunk 1 above)
Comment (gentoofreak) [inline nixos/modules/config/swap.nix]: making the cipher enumerable / computable would make this option much more complex because there are lots of different possible values (and those may change over time when cryptsetup gets updates), even though only a few of them are commonly used.

just to give some common examples:
- aes-xts-plain64
- serpent-xts-plain64
- twofish-xts-plain64
- aes-cbc-essiv:sha256
- serpent-cbc-essiv:sha256
- twofish-cbc-essiv:sha256

so basically it is `(aes|serpent|twofish)-(cbc|xts)-(plain64|essiv:[hashalgo])` and depending on cbc or xts the key size varies.

CBC: 128 or 256 bit
XTS: 256 or 512 bit

And depending on compile- and runtime options the available hashing algorithms are SHA1, SHA256, SHA512, ripemd160 and whirlpool

**Review by fpletz — APPROVED** [review]

**Comment by abbradar** [comment]
@gentoofreak Sorry, I had a very hectic month along with vacation and forgot about this.

Looking at this now 10c6df2e3c2b9d208071447bcd76e4e28e4e12dc was commited since that time. It looks like a lite version of your patch. Can you rebase atop of it? I promise I'll see it promptly this time ~_~

**Review by globin — APPROVED** [review]

**Comment by gentoofreak** [comment]
@abbradar ok, I rebased my branch... but somehow git did weird things so I see my commits twice. I'm not sure how to fix this, but it's working anyway.

**Comment by abbradar** [comment]
Tested in a local VM; it works fine for me. Thank you! I've rebased your commits into c3d5cfdc3ca709a9c5081b1a11bca533bc4788af and 9be26f81caf167f32c6af5cb876f48d7fad7d2c3.
